tests/compiler/test_relations_inverse.py

```diff
@@ -0,0 +1,300 @@
+"""
+    Copyright 2022 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import contextlib
+import textwrap
+
+import pytest
+
+from inmanta import compiler
+from inmanta.ast.statements.generator import IndexAttributeMissingInConstructorException
+
+
+def test_relations_implicit_inverse_simple(snippetcompiler) -> None:
+    """
+    Verify that the basics of implicit inverse relations on index attributes work: if an index attribute is missing in a
+    constructor call attempt to derive it from the left hand side.
+    """
+    snippetcompiler.setup_for_snippet(
+        """
+        entity A: end
+        entity B: end
+        implement A using std::none
+        implement B using std::none
+
+        A.b [0:1] -- B.a [1]
+
+        index B(a)
+
+        # nested constructors
+        a1 = A(b=B())
+        # attribute assignment
+        a2 = A()
+        a2.b = B()
+
+        assert = true
+        assert = a1.b.a == a1
+        assert = a2.b.a == a2
+        """
+    )
+    compiler.do_compile()
+
+
+def test_relations_implicit_inverse_composite_rhs(snippetcompiler) -> None:
+    """
+    Verify that implicit inverse relations on index attributes work if the constructor appears as an element in a composite
+    statement, e.g. a list or a conditional expression.
+    """
+    snippetcompiler.setup_for_snippet(
+        """
+        entity A: end
+        entity B: end
+        implement A using std::none
+        implement B using std::none
+
+        A.b [0:] -- B.a [1]
+
+        # not very realistic model, but still a good case to check
+        index B(a)
+
+        # nested constructors
+        a1 = A(b=[B(), true ? B() : B()])
+        # attribute assignment
+        a2 = A()
+        a2.b = false ? [B(), B()] : [B(), B(), B()]
```

Comment:
Those will all collapse into a single entity, maybe add another test where the index is composed of the `a` and another attribute?  So that
1. A more complex index is tried
2. We have more than one item in the relation that gets assigned the bidirectional index value

Reply:
Very good catch, all my indexes so far have only one attribute but I'm pretty sure my change should work for larger indexes as well. Will fix!

Reply:
Added a simple test case and modified this one to use the composite index.